### PR TITLE
Generalized PointCloud

### PR DESCRIPTION
**Pull request by hubernikus**
# Problem
I've been playing around with pybind / pcl. And realized a few things:

- PCL is highly templated. Unfortunately, that can not easily be translated to python using nanobind (https://nanobind.readthedocs.io/en/latest/classes.html#binding-classes-with-template-parameters). The option of appending the point-type to the corresponding point cloud / procedure. However, that has few downsides: it leads to rather long (typo-prone) names, the user needs to know exactly what point type is being used and each function type hint (in user-python code) has to be adapted for this. 

- Point clouds are big - python is slow: From my experience working with point clouds, I often have >100k points. Replicating the implementation of pcl in python, will require many loops to assign / obtain / modify points in a point cloud in python. This leads to slow code, which make the bindings potentially hard to use.

- Passing an argument to be populated is very uncommon in python, i.e. in `NormalEstimation` the output has to be past as an argument, see https://github.com/hubernikus/pcl-python-bindings/blob/1a635b5ac95e66eb084f79b56d471cde062fa1a8/src/python_features.cpp#L67
While this is very common in cpp, this pattern is very uncommon in python, where the user does not care about ownership.

- No keyword-argument in cpp: Since cpp does not have keyword arguments, function headers are kept short with focus on overloading. The beauty of keyword arguments is that the user of a library decides the order and which ones should be left to default. This makes it much easier have multiple arguments in a single function header.

# Contribution / Approach
I tried an approach which differs more from the cpp-pcl implementation, but make the library more intuitive to use for a python user. 

### Everything is a PointCloud (Tensor Interface)
 In my opinion, it is rare that a user needs to access a single point only. Most the time PointClouds have >10k points, which need to be ALL processed, filtered, visualized etc. Hence, exposing PoinTypes adds little value. Rather, the PointCloud take a specific type at construction, and handle the individual points themselves. 
The interface is with getter/setter to obtain all values (position, normal, etc) as a numpy array. Using this arrays also allows for efficient further processing.
- PointType-Enum which is passed to PointCloud 

### Hide Templatization
Let the compiler choose and perform the correct template, without having the user worry about the matching.
I think this will make the code much cleaner, and changing PointType in existing code will be possible without having to modify each function. 
- std::variant to hold point cloud (pointers). This allows automatic matching of functions.
- Create small wrapper-classes for input / output values of PointCloud modifiers.

### Many Constructor Arguments / Expose Attribute (Setter / Getter)
Pythonic function headers with all parametrization arguments. They can be set or omitted, since defaults values exist.
Additionally, the attributes are exposed (using the cpp getter/setter functions), but as attribute in the python classes.

### Chose Default
Instead of setting input/output clouds which are later modified, I think it's more understandable python code if a default behavior is provided, e.g., 
NormalEstimation -> directly assign the values to the inputPointCloud, since they values are strongly connected to the points
Filter -> create and return a new PointCloud to keep the original one

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -57,32 +59,46 @@ find_package(nanobind CONFIG REQUIRED)
 #  src/nanobind_example_ext.cpp
 #)
 
-nanobind_add_module(pcl_common_ext src/python_common.cpp)
-target_link_libraries(pcl_common_ext PRIVATE pcl_common)
+# Setup for linux, see information on
```
Comment (hubernikus) [inline CMakeLists.txt:62]: I got errors with the original linking in the ubuntu-docker container. Following this setup resolved that.

Comment (mvieth) [reply CMakeLists.txt:62]: What was the linking error exactly? The approach with setting INSTALL_RPATH to /usr/local/lib only works on Linux (maybe macOS?) and probably only if the PCL libraries are installed there?

**Comment by hubernikus** [comment]
> If I interpret your test code correctly, your suggestion would replace cloud = PointcloudXYZ() by cloud = PointCloud(t.PointXYZ)? I am not sure I see the advantage. I did not look at type hints with nanobind yet, so I cannot comment on that at the moment.

My reasoning behind was to have a single `PointCloud` class. Hence there is no class/operator overloading needed for each different PointCloud-type, but rather the evaluation happens automatically. 

> I agree that the python bindings have to be competitive in terms of speed. However, I am not sure I understand your argument about loops to assign/obtain/modify - I would expect that Python users will only have to touch the points inside a point cloud object very rarely, the actual processing all happens on the C++ side, in the templated PCL classes. Yes, the current tests add points via append in a loop, but that is more for testing purposes, and I would expect that in real use, the point clouds come from files or from sensors. 

I agree that points will mostly come from file or sensors, but from personal experience, there is often some processing needed. A larger part might be covered by the existing filtering-methods,  but I'm not sure if all are, e.g., removing points which are at a certain distance. 
However, a tensor interface could also be just on top of the standard interface. 

> I should also add that we have not optimized the PointcloudXYZ bindings for speed so far, there are probably some options we did not use yet.

I think even with optimization the difference will show similarly, since python loops are just slow and to be avoided. Any reasonably fast python code heavily depends on optimized libraries (with underlying cpp / c code), e.g., numpy & scipy for computations. 

> If I see it correctly, one major part of your proposal is that it is no longer necessary to specify the point types for each class, e.g. one can use PassThrough instead of PassThroughXYZ, PassThroughXYZRGBA etc. 

Yes that is the main reason behind doing the abstraction. This esnures that any new point-type which is added does not require adapting of the methods / classes. It is automatically enabled and type-checked through the cpp `requires`.

> While it is nice to not be required to specify the point type again and again, this approach makes a lot of extra C++ code necessary (class PassThrough, filter_return, etc), and (what concerns me the most) to duplicate all the members that the templated PCL classes already have (min, max, field_name, negative, etc).

I agree, the additional code and double and the duplicated classes are a large downside of the proposed implementation. That's the best I came up with so far.

On the other hand, using proposed approach automatically allows to use all point-types with a single class only. Whereas for the multi-class approach, there is duplication since each different point-type requires binding a new cpp-class.

In my opinion, if the goal is to expose many of the point types, using the additional abstraction will actually lead to less duplication. 
If there will only be very few point types exposed (1 or 2), then the abstraction is not worth it. 

> One alternative idea we could discuss (haven't really tested it, but should require far less extra code):

That could be nice, since a member methods would help a user choose automatically use the correct class / factory.
But it also deviates from the cpp code, and would still require duplicates for each point-type, too

> In general, we have to find a balance between what feels natural in Python, and how the PCL C++ interface does it. Of course the bindings should not feel weird to use in Python, but I think making them completely different from the C++ way is also a bad idea because it would feel very unfamiliar for people who are already using PCL in C++, and it would make things like prototyping in Python then moving to C++ quite difficult. 

I agree that changing as close as possible to the cpp implementation will make it much easier to maintain the bindings in the future. 

> I would definitely like to get the opinions of the PCL community regarding what the Python interface should look like (what they want to be able to do etc) before making a final decision.

From my experience, I've been using open3d and the non-official pcl-bindings in python.
I used the tensor interface of open3d, and it helps a lot at doing custom processing / filtering, e.g., moving some points or having a position dependent filter. 
Sometimes it is just for prototyping, at other times we leave the python code since with tensor interface / operation, the computation is fast enough. 
Additionally,  a tensor interface for the point clouds is needed for machine learning libraries (tensorflow / pytorch / skicit-learn), which is to my knowledge state-of-the-art for computer vision.

**Comment by hubernikus** [comment]
I added all the Points / Normals (as to my understanding). 
The existing procedures (features / filters / registration / segmentation) directly work with the added points, too.

What is still missing is the setter / geter for the new attributes (curvature, color, etc.)
